kratos/containers/dynamic_dimensional_array.h

```diff
@@ -0,0 +1,253 @@
+
+    private:
+        ///@name Private member variables
+        ///@{
+
+        const bool mIsManaged;
+
+        TDataType * mpData = nullptr;
+
+        ///@}
+        ///@name Private operations
+        ///@{
+
+        //*********************************************
+        //this block is needed for refcounting
+        mutable std::atomic<int> mReferenceCounter{0};
+
+        friend void intrusive_ptr_add_ref(const PointerWrapper* x)
+        {
+            x->mReferenceCounter.fetch_add(1, std::memory_order_relaxed);
+        }
+
+        friend void intrusive_ptr_release(const PointerWrapper* x)
+        {
+            if (x->mReferenceCounter.fetch_sub(1, std::memory_order_release) == 1) {
+            std::atomic_thread_fence(std::memory_order_acquire);
+            delete x;
+            }
+        }
+        //*********************************************
+
+        ///@}
+    };
+
+    ///@}
+    ///@name Type definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(DynamicDimensionalArray);
+
+    using IndexType = std::size_t;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    /**
+     * @brief Construct a new data DynamicDimensionalArray with a given a dynamic dimensional array @p rShape
+     * @details This constructor will only allocate memory for the given shape.
+     * @warning The values will be uninitialized.
+     *
+     * @param rShape Dynamic Dimensional Array shape.
+     */
+    DynamicDimensionalArray(
+        const DenseVector<unsigned int>& rShape);
+
+    /**
+     * @brief Construct a new Dynamic Dimensional Array with a given @p rShape, and initialized with the @p Value.
+     *
+     * @param rShape Dynamic Dimensional Array shape.
+     * @param Value  Initialization value.
+     */
+    DynamicDimensionalArray(
+        const DenseVector<unsigned int>& rShape,
+        const TDataType Value);
+
+    /**
+     * @brief Construct a new Dynamic Dimensional Array with given data pointer, shape.
+     * @details This constructor will construct a dynamic dimensional array with the @p pData pointer
+     *          and the given shape.
+     *              - If @p Copy = true, then the data in the @p pData will be copied
+     *              - If @p Copy = false, then the internal data will point to the @p pData .
+     *
+     * @warning This may SEGFAULT if a @ref DynamicDimensionalArray is created with @p Copy = false, and when
+     *          any of the @ref ViewData method is called while the @p pData is deallocated.
+     *
+     * @param pData     Pointer to the data
+     * @param rShape    Dynamic Dimensional Array shape.
+     * @param Copy      Whether to copy the data @p pData referring to or not.
+     */
+    DynamicDimensionalArray(
+        TDataType * pData,
+        const DenseVector<unsigned int>& rShape,
+        const bool Copy = true);
+
+    /**
+     * @brief Construct a new Dynamic Dimensional Array with @p rShape and by always copying the data pointer by @p pData
```

Comment:
I guess you mean copying the array pointed to by `pData` instead of just copying the pointer.

```suggestion
     * @brief Construct a new instance with @p rShape, copying the array referenced by @p pData.
```